RFC: Drop the user and group identifiers from the docker build

thebigcorporation opened the pull request:
- this eliminates per-user images for otherwise identical containers
 - the docker run --user parameter will substitute for this
 - security is not impacted because trivial hack-arounds remain, that are generally addressed by Apptainer

kscott-1 commented:
Pulled this down - I like it, scripts are cleaner and reduces bloat. The only pause I have here is the fact that it appears I can overwrite any container that you create with ease. What happens if you send a pipeline through with a specific 'R' container in mind that has something like GENESIS installed, but then another user is running some tests and wants to create an 'R' container with something else. The original hihg-um/R then gets killed and the pipeline will end up failing.

This isn't a _huge_ issue per se, I think there are some solutions to be had, but we'll have to think on that. Is there a way to create a base container for each repo and then lock it from being overwritten? Any extra containers will need alternate tags or something like that..

This method also appears to attach a user id and group id to the user, so the container itself has no information on the user name or group name

thebigcorporation commented:
> Pulled this down - I like it, scripts are cleaner and reduces bloat. The only pause I have here is the fact that it appears I can overwrite any container that you create with ease. What happens if you send a pipeline through with a specific 'R' container in mind that has something like GENESIS installed, but then another user is running some tests and wants to create an 'R' container with something else. The original hihg-um/R then gets killed and the pipeline will end up failing.
> 
> This isn't a _huge_ issue per se, I think there are some solutions to be had, but we'll have to think on that. Is there a way to create a base container for each repo and then lock it from being overwritten? Any extra containers will need alternate tags or something like that..
> 
> This method also appears to attach a user id and group id to the user, so the container itself has no information on the user name or group name

A running container would never be deleted or overwritten, even if changes are made.
It would simply create a new overlay from the highest common overlay.

A tag, like latest, however, could be moved, so another invocation of :latest would indeed create the scenario you describe.

I am envisioning attaching the short git  sea with --dirty as applicable to any build, so every container would have a different version; e.g. r:148bb9c-dirty would tell you this is at commit 148bb9c and has non-committed (red flag) changes.

thebigcorporation committed:
Drop the hard-coded user and root configuration for containers

 - this enables external configuration and simplifies the build
 - the implementation can serve as reference for updating other
   containers

Signed-off-by: Sven-Thorsten Dietrich <sxd1425@miami.edu>

thebigcorporation committed:
Remove the automatic :latest tag

 - additional logic is required to determine whether the upstream
   git branch is consistent with the code being built, and only then
   a :latest tag should be applied. This would usually only be done
   by automation.
 - for reproducible analyses, the code should build a clean container
   that is not flagged as -dirty. That implies, that there are no
   uncommitted changes and no untraced files, in the clean build
   directory.

Signed-off-by: Sven-Thorsten Dietrich <sxd1425@miami.edu>

kscott-1 commented:
Looks good. I'll pull this down.

thebigcorporation committed:
Alphabetize target lists

Signed-off-by: Sven-Thorsten Dietrich <sxd1425@miami.edu>
